Generic socket CAN support

jukkar opened the pull request:
This is a rewrite of #11454. It builts on top of raw socket support #12564 and virtual socket layer.
~~Please ignore the raw socket patches in this pr and if you have comments to it, go directly to #12564~~.

The BSD socket CAN support works the same way as in Linux.

The patches that are part of the socket CAN:
```
net: link: Add link addres type for CANBUS addresses
net: socket: can: Add socket CAN support
net: l2: Add CANBUS L2 layer
net: can: Add CAN handling to net_context
```

jukkar committed:
net: Add CAN protocol family type used in socket CAN support

This commit adds PF_CAN and AF_CAN protocol family identifiers
that are used by BSD socket CAN support code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>
Signed-off-by: Vipin Anand <vipin.anand@intel.com>

jukkar committed:
net: socket: can: Add socket CAN support

This allows user to create a CAN socket and to read/write data
from it. From the user point of view, the BSD socket CAN support
works same way as in Linux.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: l2: Add CANBUS L2 layer

This is basically a dummy layer that just passes data through.
It is needed so that we can create CANBUS type network interface
to the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: can: Add CAN handling to net_context

Make sure that we can work with CANBUS based sockets
in net_context API.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: link: Add link addres type for CANBUS addresses

Needed by CAN driver when socket API is enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

alexanderwachter commented on include/can.h:
```diff
@@ -96,6 +109,16 @@ enum can_mode {
+ * bit 31   : frame format flag (0 = standard 11 bit, 1 = extended 29 bit)
+ */
+typedef u32_t canid_t;
+
```
IMO its confusing to have two types of identifier now. canid_t and the can_msg itself. Maybe its better to offload the socketcan things to a socketcan.h where you include can.h?

alexanderwachter commented on drivers/can/stm32_can.c:
```diff
@@ -888,4 +890,71 @@ static void config_can_1_irq(CAN_TypeDef *can)
+
+	return ret;
+}
+
```
socket_can_send and socket_can_iface_init looks very generic. Could this be offloaded to socketcan.c ?
Maybe we get more than one driver implementation in future :-)

jukkar replied:
Yes, that can be done.

alexanderwachter commented:
Any ideas how you will do the receiving part? Especially the filter configuration?
My suggestion is to attach_isr and use the same isr callback for every filter.

jukkar commented:
I moved driver and sample code into a separate PR #12808 in order to speed up merging the socket CAN support. This PR contains now the generic code to support socket CAN.

pfalcon commented on include/net/net_ip.h:
```diff
@@ -40,12 +40,14 @@ extern "C" {
 #define PF_UNSPEC	0	/* Unspecified.  */
 #define PF_INET		2	/* IP protocol family.  */
 #define PF_INET6	10	/* IP version 6.  */
+#define PF_CAN		13	/* CAN protocol family */
```
Any hint to add to commit message where the exact number comes from?

jukkar replied:
done

pfalcon commented on include/net/net_ip.h:
```diff
@@ -157,6 +157,12 @@ struct sockaddr_ll_ptr {
 	u8_t        *sll_addr;    /* Physical-layer address  */
 };
 
+struct sockaddr_can_ptr {
```
Any hint to add to commit message and the PR description whether this follows some well-known CAN Socket API. E.g., Linux'? If no, the suggestion is to follow up Linux, up to providing the sample which shares the code and works the same way on Linux vs Zephyr.

jukkar replied:
done

pfalcon replied:
> From user point of view, the BSD socket CAN support
works same way as in Linux.

;-) Well, I guess it's better than nothing ;-).

pfalcon reviewed (commented):
The usual request to provide references of what existing de jure/de fact standards this API is based. (I'm sorry if I missed it; I look at the most obvious places. Arguably, that's the most important information about this patch.)

jukkar commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -0,0 +1,362 @@
+
+extern const struct socket_op_vtable sock_fd_op_vtable;
+
+static const struct socket_op_vtable can_sock_fd_op_vtable;
```
Just following the convention here, the tls and raw socket support is done same way.

pfalcon replied:
> why couldn't this can specific stuff go directly into socket.c ?

The whole idea is to avoid patching plain socket.c on adding each new socket type. It still need to be patched to overload constructor (i.e. socket()), and we need to implement a solution for that.

alexanderwachter replied:
Remark: I'm currently working on ISO-TP for zephyr. Maybe you should keep in mind that RAW CAN is not the only protocol for CAN.

jukkar replied:
> Remark: I'm currently working on ISO-TP for zephyr.

Good to know. I actually removed the ISO-TP address from the sockaddr_can struct as there was no use for it in this PR.

tbursztyka replied:
@pfalcon @jukka tls in itself is so specific that it indeed makes sense to split. But raw socket and can? 95% of the code is plain same. Unless we don't give a damn about small targets...

alexanderwachter replied:
But this will take some time. I have to touch the stm23 CAN driver because it can happen that data is sent out of order. This is because the mailbox sending priority is handled by hardware. This probably happens to the can raw socket too if you fill up all mailboxes without waiting (can_send callback is not NULL).

pfalcon replied:
@tbursztyka 

> 95% of the code is plain same. 

I so far haven't seen this (need to review in more detail). As a pre-statement, I seriously doubt that there's much in common in recv() implementation for TCP and RAW. Maybe there's more in common of recv() implementation for UDP and RAW. We certainly should make reusable code to be actually reusable. We probably can merge handling of some different socket types into single implementation (like is currently done with TCP and UDP). All that can be premature optimization, or artifact of subideal implementation. For example, I can't imagine CAN and normal IP to have much in common re: implementation.

Anyway, again, let me have a deeper look first. And we bigger problem actually: lack of clear understanding of what's actually being implemented. (I generalize that even to authors of PRs, because they can't clearly convey that information, which I treat as a sign of not clear enough understanding, sorry).

pfalcon commented on include/net/net_ip.h:
```diff
@@ -41,12 +41,14 @@ extern "C" {
 #define PF_INET		2	/* IP protocol family.  */
 #define PF_INET6	10	/* IP version 6.  */
 #define PF_PACKET	17	/* Packet family. */
+#define PF_CAN		29	/* Controller Area Network.  */
```
> In fact we don't care

As you can see, some people care ;-)

dbkinder commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+Overview
+********
+
+This sample is a simple RAW socket application showing usage of
```
Ethernet is a registered trademark so it's always capitalized.

Let's change this paragraph to fix grammar and spelling:
```
This sample is a simple application showing usage of
raw sockets over Ethernet.  This sample prints every Ethernet
packet received and sent over the wire. The Zephyr
network subsystem does not touch any of the headers (e.g., L2, L3). 
```

dbkinder commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+********************
+
+When the application runs, it opens a raw socket and prints
+the length of packet it receives. And sends dummy packet over
```
```suggestion
the length of each packet it receives. It also sends a dummy packet over
```

dbkinder commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+
+When the application runs, it opens a raw socket and prints
+the length of packet it receives. And sends dummy packet over
+the wire at a rate every 5 seconds.
```
```suggestion
the wire every 5 seconds.
```

dbkinder commented on samples/net/sockets/raw/README.rst:
```diff
@@ -0,0 +1,27 @@
+the length of packet it receives. And sends dummy packet over
+the wire at a rate every 5 seconds.
+
+See the `net-tools`_ project for more details.
```
more details of what?  raw sockets?

jukkar committed:
net: context: Use signed value for interface index

As the value 0 is a valid network interface index, we cannot use
unsigned value for interface index as that would not allow to
distinguish an invalid value. So make interface index a signed
8-bit value which is ok as we do not expect to have more than 127
network interfaces in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: socket: can: Add getsockopt() and setsockopt() support

It is possible to set the filter in user application and that
information is passed to the CANBUS device driver.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Various updates and setsockopt/getsockopt support added.

alexanderwachter commented:
Where do you loop-back messages that match the attached filters? I can't find this in here.
According to the linux implementation all threads that bind should see the messages as they were connected to the bus externally. I think this should somehow mimic the behavior of AUTOSAR

jukkar committed:
tests: net: context: Fix invalid address family test

The test had hard coded address family with value 4. This test
was failing as we have now AF_CAN family with value 4. Use 99
as an invalid address family value instead.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -0,0 +1,365 @@
+		return -1;
+	}
+
+	ret = net_context_sendto(pkt, dest_addr, addrlen, NULL, timeout, NULL,
```
you will need to switch to net_context_sendto_new() so you won't build the pkt here in socket, but net_context will

jukkar replied:
I was hoping we could postpone this change to bit later, after all the net_context_sendto() is still there and is working just fine.

tbursztyka replied:
As you like, but check anyway: I think it is not too much work to switch to net_context_sendto_new , if that's so: please switch.

tbursztyka commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -0,0 +1,365 @@
+		recv_len = max_len;
+	}
+
+	net_frag_linearize(buf, recv_len, pkt, 0, recv_len);
```
Use the new api: net_pkt_read_new()

jukkar commented:
> Where do you loop-back messages that match the attached filters? I can't find this in here.
> According to the linux implementation all threads that bind should see the messages as they were connected to the bus externally. I think this should somehow mimic the behavior of AUTOSAR

Can you elaborate how the system should work here? I thought the controller would send messages back to us if needed. If this is not the case, then we need to obviously handle that case.

AUTOSAR (AUTomotive Open System ARchitecture) seems to be a middleware for applications, do you have plans to do something in this front?

alexanderwachter reviewed (commented):
> > Where do you loop-back messages that match the attached filters? I can't find this in here.
> > According to the linux implementation all threads that bind should see the messages as they were connected to the bus externally. I think this should somehow mimic the behavior of AUTOSAR
> 
> Can you elaborate how the system should work here?

https://www.kernel.org/doc/Documentation/networking/can.txt
Take a look at 3.2
> I thought the controller would send messages back to us if needed. If this is not the case, then we need to obviously handle that case.

Unfortunately driver loopback mode does not work here because it disconnect the RX line.
Messages that are sent don't appear in the rx buffer.
I'm afraid that we need so store a copy of every attached filters and check them when we send anything to the bus and dispatch those messages to the bound threads.
> 
> AUTOSAR (AUTomotive Open System ARchitecture) seems to be a middleware for applications, do you have plans to do something in this front?

No. definitely not. AUTOSAR is in my opinion a standard that should be replaced by something more up to date. This is just a guess why the linux driver is implemented like hat.

jukkar commented:
> Unfortunately driver loopback mode does not work here because it disconnect the RX line.
> Messages that are sent don't appear in the rx buffer.
> I'm afraid that we need so store a copy of every attached filters and check them when we send anything to the bus and dispatch those messages to the bound threads.

Ok, makes sense. I just wonder should the message to be sent be checked in actual CAN send function like can_stm32_send().

alexanderwachter commented:
I would check them in the socket can implementation. For normal can it absolutely makes no sense to loopback messages we just sent.

jukkar commented:
Use the new net_pkt api when sending and receiving.

alexanderwachter commented:
>Sorry about these stupid newbie questions, the CANBUS is a new thing to me.
No problem. I'm not 100% an expert too ;-)

jukkar committed:
net: utils: Fix net_proto2str() function

The function did not take the address family into account
when printing protocol name. The protocol value depends on
address family.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
Fixed the code so that sending using send() and receiving using recvfrom() works.

tbursztyka commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -0,0 +1,353 @@
+
+	if (!(flags & ZSOCK_MSG_PEEK)) {
+		net_pkt_unref(pkt);
+	}
```
else { net_pkt_cursor_init(pkt); }  so another msg peek on the same pkt will work

pfalcon commented:
Wasn't a merge here a bit too hasty?

pfalcon commented on include/net/net_linkaddr.h:
```diff
@@ -45,6 +45,7 @@ enum net_link_type {
 	NET_LINK_BLUETOOTH,
 	NET_LINK_ETHERNET,
 	NET_LINK_DUMMY,
+	NET_LINK_CANBUS,
```
I thought NET_LINK_DUMMY was supposed to be last.

jukkar replied:
Should not matter, number is just a number.

pfalcon commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -108,6 +108,12 @@ config NET_SOCKETS_PACKET
 
+config NET_SOCKETS_CAN
+	bool "Enable socket CAN support [EXPERIMENTAL]"
+	default n
```
@ulfalizer asked to not use "default n", which is default itself.

jukkar replied:
Yes, this can be fixed.

pfalcon commented on subsys/net/lib/sockets/Kconfig:
```diff
@@ -108,6 +108,12 @@ config NET_SOCKETS_PACKET
+	bool "Enable socket CAN support [EXPERIMENTAL]"
+	default n
+	help
+	  The value depends on your network needs.
```
> The value depends on your network needs.

Really? Good description, can use it anywhere ;-).

jukkar replied:
You are welcome.

pfalcon commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -129,6 +129,12 @@ int _impl_zsock_socket(int family, int type, int proto)
 #endif
 
+#if defined(CONFIG_NET_SOCKETS_CAN)
+	if (family == AF_CAN && type == SOCK_RAW) {
```
This is the same problem as with initial packet socket patch. Please dispatch the entire address family to your implementation, and deal with different socket types within it there. Please fix this in https://github.com/zephyrproject-rtos/zephyr/pull/12808

jukkar replied:
I have actually a new PR coming that will remove all these settings from sockets.c. So socket families will just register their stuff in their individual files and sockets.c will use them automatically. I will send the PR next week.

pfalcon replied:
Cool. We'll have it under review. In the meantime, please fix this issue in https://github.com/zephyrproject-rtos/zephyr/pull/13168

pfalcon commented on subsys/net/lib/sockets/sockets_can.c:
```diff
@@ -0,0 +1,355 @@
+ */
+static int can_sock_listen_vmeth(void *obj, int backlog)
+{
+	return 0;
```
Same problem as with packet sockets - it's error to call these functions on non-SOCK_STREAM.

pfalcon commented on include/net/socket_can.h:
```diff
@@ -47,6 +48,20 @@ struct sockaddr_can {
+/**
+ * CAN L2 driver API. Used by socket CAN.
+ */
+struct canbus_api {
```
This probably should be called canbus_driver_api.

alexanderwachter replied:
This is not the driver. This is just a wrapper.

pfalcon replied:
> This is not the driver. This is just a wrapper.

Then it's confusing and likely unneeded wrapper. A lot in this PR seems to be done to twist net/ip/ to push unrelated protocol thru it.

pfalcon commented on subsys/net/l2/canbus/canbus.c:
```diff
@@ -0,0 +1,45 @@
+{
+	net_pkt_lladdr_src(pkt)->addr = NULL;
+	net_pkt_lladdr_src(pkt)->len = 0;
+	net_pkt_lladdr_src(pkt)->type = NET_LINK_DUMMY;
```
So, why NET_LINK_CANBUS was introduced, if NET_LINK_DUMMY is used here anyway?

jukkar replied:
Bug.

pfalcon commented on include/net/net_context.h:
```diff
@@ -380,10 +380,10 @@ static inline void net_context_set_family(struct net_context *context,
-	    family == AF_INET6 || family == AF_PACKET) {
-		/* Family is in BIT(4) and BIT(5)*/
-		flag = family << 4;
+	if (family == AF_UNSPEC || family == AF_INET || family == AF_INET6 ||
```
This is growing unwieldy. We should learn to use '>=` and `<=` math operators.

pfalcon commented:
So, there're indeed cleanups which should have been, and still should be made to this patch.

And just as with the initial patch, I don't understand why it was needed to route CAN operations thru net/ip/ (IP, **IP!!!1**). Adding unrelated protocols turns **IP** stack code into a mess.

pfalcon commented on subsys/net/ip/net_context.c:
```diff
@@ -133,24 +134,32 @@ int net_context_get(sa_family_t family,
+			if (family != AF_PACKET && family != AF_CAN) {
+				return -EINVAL;
+			}
+		} else if (IS_ENABLED(CONFIG_NET_SOCKETS_RAW) &&
```
@jukkar: Another issue with this patch. As @galak points out, CONFIG_NET_SOCKETS_RAW is dead symbol. Well, I don't understand this "if" staircase which you with @rveerama1 added here, so I hope you guys can look into what would be the proper fix.
